Issue #4533 Hard close from Dispatcher

gregw opened the pull request:
#4533 Do hard close from Dispatcher so response wrappers may intercept close.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4533 Hard close from Dispatcher

Signed-off-by: Greg Wilkins <gregw@webtide.com>

#4533 Do hard close from Dispatcher so response wrappers may intercept close.

sbordet commented on jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java:
```diff
@@ -373,6 +377,98 @@ public void testForwardFilterToRogerServlet() throws Exception
+                @Override
+                public void close() throws IOException
+                {
+                    buffer.close();
```
I have commented out this line `buffer.close()`, and the test still passes.

I'm not sure at this point what are we really testing? I would have expected the test to fail if `buffer.close()` was commented out?

If the point of the test is to verify that the wrapped `close()` is invoked, can you make it explicit with a `boolean` or a `CountDownLatch`?

gregw replied:
If you remove the method or call super.close() then the test fails.   The test correctly reproduces the failure produced by the client when the fix is not in place.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Dispatcher.java:
```diff
@@ -219,8 +219,18 @@ protected void forward(ServletRequest request, ServletResponse response, Dispatc
+                    {
+                        response.getWriter().close();
+                    }
+                }
```
There remains only one other usage of `softClose()`. Is it really necessary, or `softClose()` can be removed?

gregw replied:
The concept of a softClose is still needed by sendError - as we need to be able to treat the output as closed after the sendError for the remaining portion of a dispatch, only to reopen to send the actual response in the errorHandler or ERROR dispatch.    It was wrong to apply that concept to dispatch.forward, as there is no reopen semantic needed after a forward, plus there is a wrapping semantic that we broke.

gregw committed:
Issue #4533 Hard close from Dispatcher

Signed-off-by: Greg Wilkins <gregw@webtide.com>

#4533 improve test after review

gregw committed:
Issue #4533 Hard close from Dispatcher

Some renaming of methods to make it clear that softClose should only be used as part of sendError handling.  If softClose is used by other components, then sendError can be prevented from setting the error status.

Signed-off-by: Greg Wilkins <gregw@webtide.com>
